### PR TITLE
glsl-in: Perform output parameters implicit casts 

### PR DESCRIPTION
**Pull request by JCapucho**
Rant: this is incredibly cursed and I wouldn't even believe this existed if I hadn't ran into a shader that used this.

Glsl defines under `Function Definitions` (Paragraph 6.1 in glsl 4.60),
the following:

> When function calls are resolved, an exact type match for all the arguments is sought.
> (...)
> If no exact match is found, then the implicit conversions in section “Implicit Conversions” will be applied to find a match.
> (...)
> Mismatched types on output parameters (out or inout) must have a conversion from the formal parameter type to the calling argument type.

The glsl frontend wasn't performing this implicit cast for output parameters.

They are now implemented trough a proxy write (like when swizzling for output parameters), where a spill variable with the correct type is passed and on the call prologue a conversion back to the original type is performed and the value stored in the original variable.

Furthermore if the variable in declared `inout`, a conversion is also made in the call prelude from the original type to the correct type and the value is stored in the spill variable.

### CODE REVIEW

**Review by teoxoy — APPROVED** [review]
> Rant: this is incredibly cursed and I wouldn't even believe this existed if I hadn't ran into a shader that used this.

That's indeed quite some magic the GLSL spec mandates 😅

**Comment by teoxoy** [comment]
Looks like there are some test conflicts, feel free to merge once resolved.

**Review by jimblandy — CHANGES_REQUESTED** [review]

File: src/front/glsl/functions.rs [hunk 1]
```diff
@@ -739,15 +751,23 @@ impl Parser {
...
+                // The function call argument and the function definition
+                // parameter are not equal at this point, so we need to try
+                // implicit conversions.
+                //
+                // Now there are two cases, the argument is defined as a normal
+                // parameter (`in` or `const`), in this case an implicit
+                // conversion is made from the calling argument to the
+                // definition argument. If the parameter is `out` or `inout` the
+                // opposite needs to be done, so the implicit conversion is made
+                // from the definition argument to the calling argument.
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/glsl/functions.rs]: It makes sense, it also highlighted a problem where only the out conversion was being checked, it should be fixed now

**Comment by jimblandy** [comment]
It looks like the log message on commit e260a5c9 isn't right.

**Comment by JCapucho** [comment]
> It looks like the log message on commit [e260a5c](https://github.com/gfx-rs/naga/commit/e260a5c983f0ebc864a77ae3e52c00176582b2b8) isn't right.

Oh no, seems like I did something that made git override the commit message with one of another commit that was merged by master (It also seems to have squashed commits 😞)

**Comment by JCapucho** [comment]
@jimblandy should be fixed and rebased

**Review by jimblandy — APPROVED** [review]
This looks good to land.

**Pull request merged**